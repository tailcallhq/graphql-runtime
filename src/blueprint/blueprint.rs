use std::collections::hash_map::DefaultHasher;
use std::collections::{BTreeSet, HashMap};
use std::num::NonZeroU64;

use async_graphql::dynamic::{Schema, SchemaBuilder};
use async_graphql::extensions::ApolloTracing;
use async_graphql::ValidationMode;
use derive_setters::Setters;
use serde_json::Value;

use super::GlobalTimeout;
use crate::blueprint::{Server, Upstream};
use crate::lambda::{Expression, Lambda};

/// Blueprint is an intermediary representation that allows us to generate graphQL APIs.
/// It can only be generated from a valid Config.
/// It allows us to choose a different GraphQL Backend, without re-writing all orchestration logic.
/// It's not optimized for REST APIs (yet).
#[derive(Clone, Debug, Default, Setters)]
pub struct Blueprint {
    pub definitions: Vec<Definition>,
    pub schema: SchemaDefinition,
    pub server: Server,
    pub upstream: Upstream,
}

#[derive(Clone, Debug)]
pub enum Type {
    NamedType { name: String, non_null: bool },
    ListType { of_type: Box<Type>, non_null: bool },
}

impl Default for Type {
    fn default() -> Self {
        Type::NamedType { name: "JSON".to_string(), non_null: false }
    }
}

impl Type {
    pub fn name(&self) -> &str {
        match self {
            Type::NamedType { name, .. } => name,
            Type::ListType { of_type, .. } => of_type.name(),
        }
    }
    /// checks if the type is nullable
    pub fn is_nullable(&self) -> bool {
        !match self {
            Type::NamedType { non_null, .. } => *non_null,
            Type::ListType { non_null, .. } => *non_null,
        }
    }
}

#[derive(Clone, Debug)]
pub enum Definition {
    InterfaceTypeDefinition(InterfaceTypeDefinition),
    ObjectTypeDefinition(ObjectTypeDefinition),
    InputObjectTypeDefinition(InputObjectTypeDefinition),
    ScalarTypeDefinition(ScalarTypeDefinition),
    EnumTypeDefinition(EnumTypeDefinition),
    UnionTypeDefinition(UnionTypeDefinition),
}
impl Definition {
    pub fn name(&self) -> &str {
        match self {
            Definition::InterfaceTypeDefinition(def) => &def.name,
            Definition::ObjectTypeDefinition(def) => &def.name,
            Definition::InputObjectTypeDefinition(def) => &def.name,
            Definition::ScalarTypeDefinition(def) => &def.name,
            Definition::EnumTypeDefinition(def) => &def.name,
            Definition::UnionTypeDefinition(def) => &def.name,
        }
    }
}

#[derive(Clone, Debug)]
pub struct InterfaceTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
}

#[derive(Clone, Debug)]
pub struct ObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<FieldDefinition>,
    pub description: Option<String>,
    pub implements: BTreeSet<String>,
}

#[derive(Clone, Debug)]
pub struct InputObjectTypeDefinition {
    pub name: String,
    pub fields: Vec<InputFieldDefinition>,
    pub description: Option<String>,
}

#[derive(Clone, Debug)]
pub struct EnumTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub enum_values: Vec<EnumValueDefinition>,
}

#[derive(Clone, Debug)]
pub struct EnumValueDefinition {
    pub description: Option<String>,
    pub name: String,
    pub directives: Vec<Directive>,
}

#[derive(Clone, Debug, Default)]
pub struct SchemaDefinition {
    pub query: String,
    pub mutation: Option<String>,
    pub directives: Vec<Directive>,
}

#[derive(Clone, Debug)]
pub struct InputFieldDefinition {
    pub name: String,
    pub of_type: Type,
    pub default_value: Option<serde_json::Value>,
    pub description: Option<String>,
}

#[derive(Clone, Debug)]
pub struct Cache {
    pub max_age: NonZeroU64,
    pub hasher: DefaultHasher,
}

#[derive(Clone, Debug, Setters, Default)]
pub struct FieldDefinition {
    pub name: String,
    pub args: Vec<InputFieldDefinition>,
    pub of_type: Type,
    pub resolver: Option<Expression>,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub cache: Option<Cache>,
}

impl FieldDefinition {
    pub fn to_lambda(self) -> Option<Lambda<serde_json::Value>> {
        self.resolver.map(Lambda::new)
    }

    pub fn resolver_or_default(
        mut self,
        default_res: Lambda<serde_json::Value>,
        other: impl Fn(Lambda<serde_json::Value>) -> Lambda<serde_json::Value>,
    ) -> Self {
        self.resolver = match self.resolver {
            None => Some(default_res.expression),
            Some(expr) => Some(other(Lambda::new(expr)).expression),
        };
        self
    }
}

#[derive(Clone, Debug)]
pub struct Directive {
    pub name: String,
    pub arguments: HashMap<String, Value>,
    pub index: usize,
}

#[derive(Clone, Debug)]
pub struct ScalarTypeDefinition {
    pub name: String,
    pub directive: Vec<Directive>,
    pub description: Option<String>,
}

#[derive(Clone, Debug)]
pub struct UnionTypeDefinition {
    pub name: String,
    pub directives: Vec<Directive>,
    pub description: Option<String>,
    pub types: BTreeSet<String>,
}

///
/// Controls the kind of blueprint that is generated.
///
#[derive(Copy, Clone, Debug, Default)]
pub struct SchemaModifiers {
    /// If true, the generated schema will not have any resolvers.
    pub no_resolver: bool,
}

impl SchemaModifiers {
    pub fn no_resolver() -> Self {
        Self { no_resolver: true }
    }
}

impl Blueprint {
    pub fn query(&self) -> String {
        self.schema.query.clone()
    }

    pub fn mutation(&self) -> Option<String> {
        self.schema.mutation.clone()
    }

    fn drop_resolvers(mut self) -> Self {
        for def in self.definitions.iter_mut() {
            if let Definition::ObjectTypeDefinition(def) = def {
                for field in def.fields.iter_mut() {
                    field.resolver = None;
                }
            }
        }

        self
    }

    ///
    /// This function is used to generate a schema from a blueprint.
    ///
    pub fn to_schema(&self) -> Schema {
        self.to_schema_with(SchemaModifiers::default())
    }

    ///
    /// This function is used to generate a schema from a blueprint.
    /// The generated schema can be modified using the SchemaModifiers.
    ///
    pub fn to_schema_with(&self, schema_modifiers: SchemaModifiers) -> Schema {
        let blueprint = if schema_modifiers.no_resolver {
            self.clone().drop_resolvers()
        } else {
            self.clone()
        };

        let server = &blueprint.server;
        let mut schema = SchemaBuilder::from(&blueprint);

        if server.enable_apollo_tracing {
            schema = schema.extension(ApolloTracing);
        }

        if server.global_response_timeout > 0 {
            schema = schema
                .data(async_graphql::Value::from(server.global_response_timeout))
                .extension(GlobalTimeout);
        }

        if server.get_enable_query_validation() || schema_modifiers.no_resolver {
            schema = schema.validation_mode(ValidationMode::Strict);
        } else {
            schema = schema.validation_mode(ValidationMode::Fast);
        }

        if !server.get_enable_introspection() || schema_modifiers.no_resolver {
            schema = schema.disable_introspection();
        }

        // We should safely assume the blueprint is correct and,
        // generation of schema cannot fail.
        schema.finish().unwrap()
    }
}
