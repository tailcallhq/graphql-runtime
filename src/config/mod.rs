pub use config::*;
pub use key_values::*;
pub use server::*;
pub use source::*;

mod config;
mod from_document;
pub mod group_by;
mod into_document;
mod key_values;
mod n_plus_one;
pub mod reader;
mod server;
mod source;

fn is_default<T: Default + Eq>(val: &T) -> bool {
  *val == T::default()
}
