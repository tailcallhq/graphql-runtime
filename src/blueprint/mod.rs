mod blueprint;
mod compress;
mod definitions;
mod from_config;
mod into_schema;
mod mustache;
mod operation;
mod operators;
mod schema;
mod server;
mod timeout;
mod upstream;
pub use blueprint::*;
pub use definitions::*;
pub use from_config::*;
pub use operation::*;
pub use operators::*;
pub use schema::*;
pub use server::*;
pub use timeout::GlobalTimeout;
pub use upstream::*;

use crate::config::{Arg, ConfigSet, Field};
use crate::try_fold::TryFold;

pub type TryFoldConfig<'a, A> = TryFold<'a, ConfigSet, A, String>;

pub(crate) trait TypeLike {
    fn name(&self) -> &str;
    fn list(&self) -> bool;
    fn non_null(&self) -> bool;
    fn list_type_required(&self) -> bool;
}

impl TypeLike for Field {
    fn name(&self) -> &str {
        &self.type_of
    }

    fn list(&self) -> bool {
        self.list
    }

    fn non_null(&self) -> bool {
        self.required
    }

    fn list_type_required(&self) -> bool {
        self.list_type_required
    }
}

impl TypeLike for Arg {
    fn name(&self) -> &str {
        &self.type_of
    }

    fn list(&self) -> bool {
        self.list
    }

    fn non_null(&self) -> bool {
        self.required
    }

    fn list_type_required(&self) -> bool {
        false
    }
}

pub(crate) fn to_type<T>(field: &T, override_non_null: Option<bool>) -> Type
where
    T: TypeLike,
{
    let name = field.name();
    let list = field.list();
    let list_type_required = field.list_type_required();
    let non_null = if let Some(non_null) = override_non_null {
        non_null
    } else {
        field.non_null()
    };

    if list {
        Type::ListType {
            of_type: Box::new(Type::NamedType {
                name: name.to_string(),
                non_null: list_type_required,
            }),
            non_null,
        }
    } else {
        Type::NamedType { name: name.to_string(), non_null }
    }
}

pub fn is_scalar(type_name: &str) -> bool {
    ["String", "Int", "Float", "Boolean", "ID", "JSON"].contains(&type_name)
}
