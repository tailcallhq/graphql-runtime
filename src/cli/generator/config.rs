use std::env;
use std::marker::PhantomData;
use std::path::Path;

use derive_setters::Setters;
use path_clean::PathClean;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use url::Url;

use crate::core::config::{self};

#[derive(Deserialize, Serialize, Debug, Default, Setters)]
#[serde(rename_all = "camelCase")]
pub struct Config<Status = UnResolved> {
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub inputs: Vec<Input<Status>>,
    pub output: Output<Status>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub preset: Option<Preset>,
    pub schema: Schema,
}

#[derive(Clone, Deserialize, Serialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct Preset {
    merge_type: Option<f32>,
    consolidate_url: Option<f32>,
}

impl From<Preset> for config::transformer::Preset {
    fn from(val: Preset) -> Self {
        let mut preset = config::transformer::Preset::default();
        if let Some(merge_type) = val.merge_type {
            preset = preset.merge_type(merge_type);
        }

        if let Some(consolidate_url) = val.consolidate_url {
            preset = preset.consolidate_url(consolidate_url);
        }

        preset
    }
}

#[derive(Deserialize, Serialize, Debug, Default)]
#[serde(transparent)]
pub struct Location<A>(
    #[serde(skip_serializing_if = "Location::is_empty")] pub String,
    #[serde(skip)] PhantomData<A>,
);

impl<A> Location<A> {
    fn is_empty(&self) -> bool {
        self.0.is_empty()
    }
}

impl Location<UnResolved> {
    fn into_resolved(self, parent_dir: Option<&Path>) -> Location<Resolved> {
        let path = {
            let path = self.0.as_str();
            if Url::parse(path).is_ok() || Path::new(path).is_absolute() {
                path.to_string()
            } else {
                let parent_dir = parent_dir.unwrap_or(Path::new(""));
                let joined_path = parent_dir.join(path);
                if let Ok(abs_path) = std::fs::canonicalize(&joined_path) {
                    abs_path.to_string_lossy().to_string()
                } else if let Ok(cwd) = env::current_dir() {
                    cwd.join(joined_path).clean().to_string_lossy().to_string()
                } else {
                    joined_path.clean().to_string_lossy().to_string()
                }
            }
        };
        Location(path, PhantomData)
    }
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Input<Status = UnResolved> {
    #[serde(flatten)]
    pub source: Source<Status>,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Source<Status = UnResolved> {
    #[serde(rename_all = "camelCase")]
    Curl {
        src: Location<Status>,
        field_name: String,
    },
    Proto {
        src: Location<Status>,
    },
    Config {
        src: Location<Status>,
    },
}

#[derive(Deserialize, Serialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct Output<Status = UnResolved> {
    #[serde(skip_serializing_if = "Location::is_empty")]
    pub path: Location<Status>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub format: Option<config::Source>,
}

#[derive(Debug)]
pub enum Resolved {}

#[derive(Serialize, Deserialize, Debug, JsonSchema, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct UnResolved {}

#[derive(Deserialize, Serialize, Debug, Default)]
pub struct Schema {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub query: Option<String>,
}

impl Output<UnResolved> {
    pub fn resolve(self, parent_dir: Option<&Path>) -> anyhow::Result<Output<Resolved>> {
        Ok(Output {
            format: self.format,
            path: self.path.into_resolved(parent_dir),
        })
    }
}

impl Source<UnResolved> {
    pub fn resolve(self, parent_dir: Option<&Path>) -> anyhow::Result<Source<Resolved>> {
        match self {
            Source::Curl { src, field_name } => {
                let resolved_path = src.into_resolved(parent_dir);
                Ok(Source::Curl { src: resolved_path, field_name })
            }
            Source::Proto { src, .. } => {
                let resolved_path = src.into_resolved(parent_dir);
                Ok(Source::Proto { src: resolved_path })
            }
            Source::Config { src, .. } => {
                let resolved_path = src.into_resolved(parent_dir);
                Ok(Source::Config { src: resolved_path })
            }
        }
    }
}

impl Input<UnResolved> {
    pub fn resolve(self, parent_dir: Option<&Path>) -> anyhow::Result<Input<Resolved>> {
        let resolved_source = self.source.resolve(parent_dir)?;
        Ok(Input { source: resolved_source })
    }
}

impl Config {
    /// Resolves all the relative paths present inside the GeneratorConfig.
    pub fn into_resolved(self, config_path: &str) -> anyhow::Result<Config<Resolved>> {
        let parent_dir = Some(Path::new(config_path).parent().unwrap_or(Path::new("")));

        let inputs = self
            .inputs
            .into_iter()
            .map(|input| input.resolve(parent_dir))
            .collect::<anyhow::Result<Vec<Input<Resolved>>>>()?;

        let output = self.output.resolve(parent_dir)?;

        Ok(Config { inputs, output, schema: self.schema, preset: self.preset })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn location<S: AsRef<str>>(s: S) -> Location<UnResolved> {
        Location(s.as_ref().to_string(), PhantomData)
    }

    #[test]
    fn test_config_codec() {
        let config = Config::default().inputs(vec![Input {
            source: Source::Curl {
                src: location("https://example.com"),
                field_name: "test".to_string(),
            },
        }]);
        let actual = serde_json::to_string_pretty(&config).unwrap();
        insta::assert_snapshot!(actual)
    }
}
