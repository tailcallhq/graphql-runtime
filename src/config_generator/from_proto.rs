#![allow(unused)]

use std::collections::{BTreeSet, HashMap};

use convert_case::{Case, Casing};
use derive_setters::Setters;
use prost_reflect::prost_types::{
    DescriptorProto, EnumDescriptorProto, FileDescriptorSet, ServiceDescriptorProto,
};
use strum_macros::Display;

use crate::blueprint::GrpcMethod;
use crate::config::{Arg, Config, Field, Grpc, Tag, Type};

pub(super) static DEFAULT_SPECTATOR: &str = "_";

/// Contains the configuration for the config generator
pub struct ProtoGeneratorConfig {
    query: String,
}

impl ProtoGeneratorConfig {
    pub fn new(query: Option<String>) -> Self {
        Self { query: query.unwrap_or_else(|| "Query".to_string()) }
    }

    pub fn get_query(&self) -> &str {
        self.query.as_str()
    }
}

/// Enum to represent the type of the descriptor
#[derive(Display, Clone)]
pub enum DescriptorType {
    Enum,
    Message,
    Query,
}

#[derive(Default)]
struct Helper {
    map: HashMap<String, String>,
    package: String,
}

impl Helper {
    fn get_value(&self, name: &str, ty: DescriptorType) -> String {
        let package = self.package.replace('.', DEFAULT_SPECTATOR).to_uppercase();
        match ty {
            DescriptorType::Enum => {
                format!("{}{}{}", package, DEFAULT_SPECTATOR, name)
            }
            DescriptorType::Message => {
                format!("{}{}{}", package, DEFAULT_SPECTATOR, name)
            }
            DescriptorType::Query => format!(
                "{}{}{}",
                package.to_case(Case::Snake),
                DEFAULT_SPECTATOR,
                name.to_case(Case::Camel),
            ),
        }
    }
    fn insert(&mut self, name: &str, ty: DescriptorType) {
        self.map.insert(
            format!("{}.{}", self.package, name),
            self.get_value(name, ty),
        );
    }
    fn get(&self, name: &str) -> Option<String> {
        self.map.get(&format!("{}.{}", self.package, name)).cloned()
    }
}

fn convert_ty(proto_ty: &str) -> String {
    let binding = proto_ty.to_lowercase();
    let proto_ty = binding.strip_prefix("type_").unwrap_or(proto_ty);
    match proto_ty {
        "double" | "float" => "Float",
        "int32" | "int64" | "fixed32" | "fixed64" | "uint32" | "uint64" => "Int",
        "bool" => "Boolean",
        "string" | "bytes" => "String",
        x => x,
    }
    .to_string()
}

fn get_output_ty(output_ty: &str) -> (String, bool) {
    // type, required
    match output_ty {
        "google.protobuf.Empty" => {
            ("String".to_string(), false) // If it's no response is expected, we
                                          // return a nullable string type
        }
        any => (any.to_string(), true), /* Setting it not null by default. There's no way to
                                         * infer this from proto file */
    }
}

fn get_arg(input_ty: &str, helper: &mut Helper) -> Option<(String, Arg)> {
    match input_ty {
        "google.protobuf.Empty" | "" => None,
        any => {
            let key = convert_ty(any).to_case(Case::Camel);
            let val = Arg {
                type_of: helper.get(any).unwrap_or(any.to_string()),
                list: false,
                required: true,
                /* Setting it not null by default. There's no way to infer this
                 * from proto file */
                doc: None,
                modify: None,
                default_value: None,
            };

            Some((key, val))
        }
    }
}

fn get_ty(name: &str, cfg: &Config, helper: &mut Helper, ty: DescriptorType) -> Type {
    helper.insert(name, ty);
    let mut ty = cfg
        .types
        .get(&helper.get(name).unwrap())
        .cloned()
        .unwrap_or_default(); // it should be
                              // safe to call
                              // unwrap here
    ty.tag = Some(Tag { name: format!("{}.{}", helper.package, name) });
    ty
}

fn append_enums(
    config: &mut Config,
    enums: Vec<EnumDescriptorProto>,
    helper: &mut Helper,
) -> anyhow::Result<()> {
    for enum_ in enums {
        let enum_name = enum_.name();

        let mut ty = get_ty(enum_name, config, helper, DescriptorType::Enum);

        let mut variants = enum_
            .value
            .iter()
            .map(|v| v.name().to_string())
            .collect::<BTreeSet<String>>();
        if let Some(vars) = ty.variants {
            variants.extend(vars);
        }
        ty.variants = Some(variants);
        config.types.insert(helper.get(enum_name).unwrap(), ty);
        // it should be
        // safe to call
        // unwrap here
    }
    Ok(())
}

fn append_msg_type(
    config: &mut Config,
    messages: Vec<DescriptorProto>,
    helper: &mut Helper,
) -> anyhow::Result<()> {
    if messages.is_empty() {
        return Ok(());
    }
    for message in messages {
        let msg_name = message.name().to_string();

        let mut ty = get_ty(&msg_name, config, helper, DescriptorType::Message);

        append_enums(config, message.enum_type, helper)?;
        append_msg_type(config, message.nested_type, helper)?;

        for field in message.field {
            let field_name = field.name().to_string();
            let mut cfg_field = Field::default();

            let label = field.label().as_str_name().to_lowercase();
            cfg_field.list = label.contains("repeated");
            cfg_field.required = label.contains("required");

            if field.r#type.is_some() {
                let type_of = convert_ty(field.r#type().as_str_name());
                cfg_field.type_of = type_of.to_string();
            } else {
                // for non-primitive types
                let type_of = convert_ty(field.type_name());
                cfg_field.type_of = helper.get(&type_of).unwrap_or(type_of);
            }

            ty.fields.insert(field_name, cfg_field);
        }

        config.types.insert(helper.get(&msg_name).unwrap(), ty); // it should be
                                                                 // safe to call
                                                                 // unwrap here
    }
    Ok(())
}

fn generate_ty(
    config: &mut Config,
    services: Vec<ServiceDescriptorProto>,
    helper: &mut Helper,
    key: &str,
) -> anyhow::Result<Type> {
    let package = helper.package.clone();
    let mut grpc_method = GrpcMethod { package, service: "".to_string(), name: "".to_string() };
    let mut ty = config.types.get(key).cloned().unwrap_or_default();

    for service in services {
        let service_name = service.name().to_string();
        for method in &service.method {
            let method_name = method.name();

            helper.insert(method_name, DescriptorType::Query);

            let mut cfg_field = Field::default();
            if let Some((k, v)) = get_arg(method.input_type(), helper) {
                cfg_field.args.insert(k, v);
            }

            let (output_ty, required) = get_output_ty(method.output_type());
            cfg_field.type_of = helper.get(&output_ty).unwrap_or(output_ty.clone());
            cfg_field.required = required;

            grpc_method.service = service_name.clone();
            grpc_method.name = method_name.to_string();

            cfg_field.grpc = Some(Grpc {
                base_url: None,
                body: None,
                group_by: vec![],
                headers: vec![],
                method: grpc_method.to_string(),
            });
            ty.fields
                .insert(helper.get(method_name).unwrap(), cfg_field);
        }
    }
    Ok(ty)
}

fn append_query_service(
    config: &mut Config,
    services: Vec<ServiceDescriptorProto>,
    gen: &ProtoGeneratorConfig,
    helper: &mut Helper,
) -> anyhow::Result<()> {
    if services.is_empty() {
        return Ok(());
    }

    let query = gen.get_query();

    let ty = generate_ty(config, services, helper, query)?;

    if ty.ne(&Type::default()) {
        config.schema.query = Some(query.to_string());
        config.types.insert(query.to_string(), ty);
    }
    Ok(())
}

pub fn build_config(
    descriptor_sets: Vec<FileDescriptorSet>,
    gen: &ProtoGeneratorConfig,
) -> anyhow::Result<Config> {
    let mut config = Config::default();
    let mut helper = Helper::default();

    for descriptor_set in descriptor_sets {
        for file_descriptor in descriptor_set.file {
            helper.package = file_descriptor.package().to_string();

            append_enums(&mut config, file_descriptor.enum_type, &mut helper)?;
            append_msg_type(&mut config, file_descriptor.message_type, &mut helper)?;
            append_query_service(
                &mut config,
                file_descriptor.service.clone(),
                gen,
                &mut helper,
            )?;
        }
    }

    Ok(config)
}

#[cfg(test)]
mod test {

    use std::path::PathBuf;

    use prost_reflect::prost_types::{FileDescriptorProto, FileDescriptorSet};

    use crate::config_generator::from_proto::{build_config, ProtoGeneratorConfig};

    fn get_proto_file_descriptor(name: &str) -> anyhow::Result<FileDescriptorProto> {
        let mut proto_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        proto_path.push("src");
        proto_path.push("grpc");
        proto_path.push("tests");
        proto_path.push("proto");
        proto_path.push(name);
        Ok(protox_parse::parse(
            name,
            std::fs::read_to_string(proto_path)?.as_str(),
        )?)
    }

    fn get_generator_cfg() -> ProtoGeneratorConfig {
        ProtoGeneratorConfig::new(Some("Query".to_string()))
    }

    #[test]
    fn test_from_proto() -> anyhow::Result<()> {
        // news_enum.proto covers:
        // test for mutation
        // test for empty objects
        // test for optional type
        // test for enum
        // test for repeated fields
        // test for a type used as both input and output
        // test for two types having same name in different packages

        let mut set = FileDescriptorSet::default();

        let news = get_proto_file_descriptor("news_enum.proto")?;
        let greetings = get_proto_file_descriptor("greetings.proto")?;
        let greetings_dup_methods = get_proto_file_descriptor("greetings_dup_methods.proto")?;

        set.file.push(news.clone());
        set.file.push(greetings.clone());
        set.file.push(greetings_dup_methods.clone());

        let result = build_config(vec![set], &get_generator_cfg())?.to_sdl();

        insta::assert_snapshot!(result);

        // test for 2 different sets
        let mut set = FileDescriptorSet::default();
        let mut set1 = FileDescriptorSet::default();
        let mut set2 = FileDescriptorSet::default();
        set.file.push(news);
        set1.file.push(greetings);
        set2.file.push(greetings_dup_methods);

        let result_sets = build_config(vec![set, set1, set2], &get_generator_cfg())?.to_sdl();

        assert_eq!(result, result_sets);

        Ok(())
    }

    #[test]
    fn test_required_types() -> anyhow::Result<()> {
        // required fields are deprecated in proto3 (https://protobuf.dev/programming-guides/dos-donts/#add-required)
        // this example uses proto2 to test the same.
        // for proto3 it's guaranteed to have a default value (https://protobuf.dev/programming-guides/proto3/#default)
        // and our implementation (https://github.com/tailcallhq/tailcall/pull/1537) supports default values by default.
        // so we do not need to explicitly mark fields as required.

        let mut set = FileDescriptorSet::default();
        let req_proto = get_proto_file_descriptor("required_fields.proto")?;
        set.file.push(req_proto);

        let cfg = build_config(vec![set], &get_generator_cfg())?.to_sdl();
        insta::assert_snapshot!(cfg);

        Ok(())
    }
}
