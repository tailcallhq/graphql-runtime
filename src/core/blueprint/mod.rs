mod auth;
mod blueprint;
mod compress;
mod cors;
mod definitions;
mod dynamic_value;
mod from_config;
mod index;
mod into_schema;
mod links;
mod mustache;
mod operators;
mod schema;
mod server;
pub mod telemetry;
mod timeout;
mod union_resolver;
mod upstream;

pub use auth::*;
pub use blueprint::*;
pub use cors::*;
pub use definitions::*;
pub use dynamic_value::*;
pub use from_config::*;
pub use index::*;
pub use links::*;
pub use operators::*;
pub use schema::*;
pub use server::*;
pub use timeout::GlobalTimeout;
pub use upstream::*;

use crate::core::config::ConfigModule;
use crate::core::try_fold::TryFold;

pub type TryFoldConfig<'a, A> = TryFold<'a, ConfigModule, A, String>;
