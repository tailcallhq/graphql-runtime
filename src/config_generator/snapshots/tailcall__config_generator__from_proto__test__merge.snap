---
source: src/config_generator/from_proto.rs
expression: result
---
schema @server @upstream {
  query: Query
}

input Author {
  email: String
  name: String
}

input HelloRequest {
  name: String
}

input MultipleNewsId {
  ids: [NewsId]
}

input News {
  author: Author
  body: String
  foo: Status
  id: Int
  postImage: String
  title: String
}

input NewsId {
  id: Int
}

enum Status {
  DRAFT
  PUBLISHED
}

type HelloReply {
  message: String
}

type NewsList {
  news: [News]
}

type Query {
  addNews(news: News!): News! @grpc(method: "news.NewsService.AddNews")
  deleteNews(newsId: NewsId!): String @grpc(method: "news.NewsService.DeleteNews")
  editNews(news: News!): News! @grpc(method: "news.NewsService.EditNews")
  getAllNews: NewsList! @grpc(method: "news.NewsService.GetAllNews")
  getMultipleNews(multipleNewsId: MultipleNewsId!): NewsList! @grpc(method: "news.NewsService.GetMultipleNews")
  getNews(newsId: NewsId!): News! @grpc(method: "news.NewsService.GetNews")
  sayHello(helloRequest: HelloRequest!): HelloReply! @grpc(method: "greetings_dup.Greeter.SayHello")
}
